### PR TITLE
[Inductor cutlass backend] Fix cutlass_utils.get_max_alignment() for strided layouts.

### PR DESCRIPTION
**Pull request by kadeng**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #124928
* #125407
* #125406
* __->__ #124930

Fixes cutlass_utils.get_max_alignment() which was so far not checking the alignment properly. Basically
the method so far assumed that the passed layout is contiguous and row-major, which does not have to be true.

Test Plan:
CI - test_cutlass_backend.py to prevent regressions
Added unit test

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/codegen/cuda/cutlass_utils.py [hunk 1]
```diff
@@ -293,13 +293,32 @@ def get_max_alignment(inductor_layout: Layout) -> int:
...
+                        break
+                if do_continue:
+                    continue
                 return alignment
-
-    return 1
+        return 1
+    except ValueError:
+        # No dim with stride 1 found, return 1
+        return 1
```
Comment (int3) [inline torch/_inductor/codegen/cuda/cutlass_utils.py:321]: hm does the exception block need to cover all these lines, or just line 297?

File: torch/_inductor/codegen/cuda/cutlass_utils.py [hunk 2]
```diff
@@ -293,13 +293,32 @@ def get_max_alignment(inductor_layout: Layout) -> int:
...
+                for dim in range(len(size)):
+                    # All non-contiguous strides must be aligned
+                    if dim == contiguous_dim:
+                        continue
+                    dim_stride = int(inductor_layout.stride[dim])
+                    if dim_stride % alignment != 0:
+                        do_continue = True
+                        break
+                if do_continue:
+                    continue
```
Comment (int3) [inline torch/_inductor/codegen/cuda/cutlass_utils.py:316]: nit: I think an `if any(...): continue` expression here would be more readable / less imperative

File: test/inductor/test_cutlass_backend.py [hunk 3]
```diff
@@ -616,6 +617,28 @@ def select_no_algorithm(*args, **kwargs):
                             cuda_template_count += 1
                     assert cuda_template_count > 0, "No CUDATemplateCaller choices"
 
+    @unittest.skipIf(not SM80OrLater, "need sm_90")
+    @unittest.skipIf(config.is_fbcode(), "fbcode requires different CUTLASS path setup")
+    @unittest.mock.patch.dict(os.environ, {"PATH": _get_path_without_sccache()})
+    def test_get_max_alignment(self):
+        l4 = FixedLayout("cpu", torch.half, size=(1, 2, 4), stride=(0, 4, 1))
+        m4 = get_max_alignment(l4)
+        assert (
```
Comment (int3) [inline test/inductor/test_cutlass_backend.py:626]: I think python tests should use `self.assertEquals`?

**Review by int3 — APPROVED** [review]

File: test/inductor/test_cutlass_backend.py (hunk 3 above)
Comment (kadeng) [reply test/inductor/test_cutlass_backend.py]: Thanks for the review. I addressed all comments.

**Pull request closed without merging**